File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        // bottom up approach - 
        vector<int>memo(nums.size(), -1);
        return max(nums[0] + rob(nums, 2, memo), rob(nums, 1, memo));
    }
    
    int rob(vector<int>& nums, int start, vector<int>& memo) {
        if(start >= nums.size()) {
            return 0;
        }
        if(memo[start] == -1) {
            memo[start] = max(nums[start] + rob(nums, start + 2, memo),
                              rob(nums, start + 1, memo));
        }
        return memo[start];
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        memo = {}
        
        def dfs(pos, cool):
            
            if pos >= len(nums):
                return 0
            
            if cool is True:
                return dfs(pos+1, False)
            
            if pos in memo:
                return memo[pos]
            
            # steal
            op1 = dfs(pos+1, True) + nums[pos]
            # don't steal
            op2 = dfs(pos+1, False)
            
            memo[pos] = max(op1, op2)
            return memo[pos]
        
        return dfs(0, False)